lib/exq/redis/job_queue.ex

```diff
@@ -120,6 +120,59 @@ defmodule Exq.Redis.JobQueue do
     end
   end
 
+  def enqueue_all(redis, namespace, jobs) do
+    schedule_queue = scheduled_queue_key(namespace)
+
+    job_attrs = extract_job_attrs(namespace, jobs)
+
+    lock_commands =
+      job_attrs
+      |> Enum.map(fn ({ jid, _score, _job_serialized, unique_key, unlocks_in }) ->
+        if unlocks_in do
+          ["SET", unique_key, jid, "PX", unlocks_in, "NX"]
```

Comment:
The uniqueness feature can only be implemented via lua script, since we do a get, then lpush/zadd based on the get response.

I would start like this

```
MULTI // transaction
// load scripts to make sure  EVALSHA won't fail
SCRIPT LOAD // enqueue
SCRIPT LOAD // enqueue_at
EVALSHA hash key args
EVALSHA hash key args
EVALSHA hash key args
EVALSHA hash key args
EXEC
```

Reply:
Thanks for quickly reviewing @ananthakumaran

From Redix Docs:
> ## Problems with transactions

 > There's an inherent problem with Redix's architecture and MULTI/EXEC
transaction. A Redix process is a single connection to Redis that can be used
by many clients. If a client A sends MULTI and client B sends a command before
client A sends EXEC, client B's command will be part of the transaction. This
is intended behaviour, but it might not be what you expect. This is why
transaction_pipeline/3 exists: this function wraps commands in MULTI/EXEC but
sends all in a pipeline. Since everything is sent in the pipeline, it's sent at
once on the connection and no commands can end up in the middle of the
transaction.

So I had to use Redix.transaction_pipeline/3 instead of MULTI/EXEC

> since we do a get, then lpush/zadd based on the get response.

I did the same exact thing,
for each job in the List I constructed a lock aquiring redis command:

```elixir
# If job option indicates uniqueness features is wanted, this command aquires a lock:
["SET", unique_key, jid, "PX", unlocks_in, "NX"]

# If uniqueness features are not wanted,
# this command returns an "OK" indicating the lock is not already aquired
["ECHO", "OK"]
```
so this construct a List of lock aquiring commands which I am passing to Redix.transaction_pipeline to run on redis.

from there, for each job we either have: `"OK"` or `nil` which then is used to construct another list of `ZADD` to push the job if lock_response is `"OK"` or `GET` to get the conflicted jid if the lock_response is `nil`

Finally that list of commands are also passed to Redix.transaction_pipeline and from there, for each job we either construct a `{ :ok, jid }` or a `{:conflict, conflict_jid}`

I quickly tested this locally, I confirmed uniqueness features are working fine, locally. but once I get a thumbs up, I ll start writing unit tests.

Reply:
oh, wait, I just realized something. Are you saying this has to be a lua script because of the potential race condition between reading lock and enuquing the job?

Reply:
> oh, wait, I just realized something. Are you saying this has to be a lua script because of the potential race condition between reading lock and enuquing the job?

Yes, btw, I don't fully understand how your approach works, but theoretically speaking it's not going to handle the race condition involved. Check the existing implementation, it uses  [lua script](https://github.com/akira/exq/blob/master/lib/exq/redis/job_queue.ex#L41) for enqueue

You would have to figure out how to do it inside multi/exec

Reply:
> So I had to use Redix.transaction_pipeline/3 instead of MULTI/EXEC

transaction_pipeline is just `Redix.pipeline(conn, [["MULTI"]] ++ commands ++ [["EXEC"]], options)`

Reply:
Yeah i am now reading in Redis website:
> Redis guarantees the script's atomic execution. While executing the script, all server activities are blocked during its entire runtime. These semantics mean that all of the script's effects either have yet to happen or had already happened.

https://redis.io/docs/manual/programmability/eval-intro/

I ll try writing it with Lua Script :)